// ionic4-hidenav/esm5/lib/hidenav-header.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef } from '@angular/core';
import { HidenavService } from './hidenav-service.service';
var HidenavHeaderDirective = /** @class */ (function () {
    function HidenavHeaderDirective(el, globals) {
        this.el = el;
        this.globals = globals;
    }
    /**
     * @return {?}
     */
    HidenavHeaderDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var observer = new MutationObserver((/**
         * @param {?} mutations
         * @return {?}
         */
        function (mutations) {
            mutations.forEach((/**
             * @return {?}
             */
            function () {
                if (_this.el.nativeElement.getAttribute('hidenav-header').length > 0) {
                    _this.name = _this.el.nativeElement.getAttribute('hidenav-header');
                    _this.start();
                    observer.disconnect();
                }
            }));
        }));
        observer.observe(this.el.nativeElement, {
            attributes: true,
        });
    };
    /**
     * @return {?}
     */
    HidenavHeaderDirective.prototype.start = /**
     * @return {?}
     */
    function () {
        if (this.name) {
            if (typeof this.globals.data[this.name] == 'undefined' || this.globals.data[this.name] == null)
                this.globals.data[this.name] = [];
            if (this.globals.data[this.name].header != null)
                return false;
            this.globals.data[this.name].header = this.el;
            this.globals.initiate(this.name);
        }
    };
    /**
     * @return {?}
     */
    HidenavHeaderDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.name) {
            delete this.globals.data[this.name].header;
        }
    };
    HidenavHeaderDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[hidenav-header]'
                },] }
    ];
    HidenavHeaderDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: HidenavService }
    ]; };
    return HidenavHeaderDirective;
}());
export { HidenavHeaderDirective };
if (false) {
    /** @type {?} */
    HidenavHeaderDirective.prototype.name;
    /** @type {?} */
    HidenavHeaderDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    HidenavHeaderDirective.prototype.globals;
}
