// ionic4-hidenav/esm5/lib/hidenav-content.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, Host, Self, Optional, ElementRef } from '@angular/core';
import { IonContent } from '@ionic/angular';
import { HidenavService } from './hidenav-service.service';
import $ from "jquery";
var HidenavContentDirective = /** @class */ (function () {
    function HidenavContentDirective(el, contentElem, globals) {
        this.el = el;
        this.contentElem = contentElem;
        this.globals = globals;
    }
    /**
     * @return {?}
     */
    HidenavContentDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this.contentElem.nativeElement.hasAttribute('hidenav-tabspage')) {
            this.name = this.globals.requestName();
            $(this.contentElem.nativeElement).attr('hidenav-content', this.name);
            $('[hidenav-header]', $(this.contentElem.nativeElement).parents().get().find((/**
             * @param {?} itm
             * @return {?}
             */
            function (itm) { return $(itm).find('[hidenav-header]').length; }))).attr('hidenav-header', this.name);
            this.start();
        }
        else {
            /** @type {?} */
            var counter_1 = 0;
            /** @type {?} */
            var int_1 = setInterval((/**
             * @return {?}
             */
            function () {
                /** @type {?} */
                var x = $(_this.contentElem.nativeElement).closest('[hidenav-tabscontent]').attr('hidenav-tabscontent');
                counter_1++;
                if (x && x.length > 0) {
                    _this.parent = $(_this.contentElem.nativeElement).closest('[hidenav-tabscontent]').attr('hidenav-tabscontent');
                    _this.name = _this.globals.requestTabName(_this.parent);
                    $(_this.contentElem.nativeElement).attr('hidenav-content', _this.name);
                    $(_this.contentElem.nativeElement).attr('hidenav-tabspage', _this.parent);
                    _this.start();
                    clearInterval(int_1);
                }
                else if (counter_1 > 50) {
                    clearInterval(int_1);
                }
            }), 50);
        }
    };
    /**
     * @return {?}
     */
    HidenavContentDirective.prototype.start = /**
     * @return {?}
     */
    function () {
        if (this.name) {
            if (typeof this.globals.data[this.name] == 'undefined' || this.globals.data[this.name] == null)
                this.globals.data[this.name] = [];
            this.globals.data[this.name].content = this.el;
            this.globals.data[this.name].contentElem = this.contentElem;
            if (this.parent)
                this.globals.data[this.name].parent = this.parent;
            this.globals.initiate(this.name);
        }
    };
    /**
     * @return {?}
     */
    HidenavContentDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.name) {
            delete this.globals.data[this.name].content;
        }
    };
    HidenavContentDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[hidenav-content]'
                },] }
    ];
    HidenavContentDirective.ctorParameters = function () { return [
        { type: IonContent, decorators: [{ type: Host }, { type: Self }, { type: Optional }] },
        { type: ElementRef },
        { type: HidenavService }
    ]; };
    HidenavContentDirective.propDecorators = {
        hntb: [{ type: Input, args: ['hidenav-tabspage',] }]
    };
    return HidenavContentDirective;
}());
export { HidenavContentDirective };
if (false) {
    /** @type {?} */
    HidenavContentDirective.prototype.name;
    /** @type {?} */
    HidenavContentDirective.prototype.parent;
    /** @type {?} */
    HidenavContentDirective.prototype.hntb;
    /** @type {?} */
    HidenavContentDirective.prototype.el;
    /** @type {?} */
    HidenavContentDirective.prototype.contentElem;
    /**
     * @type {?}
     * @private
     */
    HidenavContentDirective.prototype.globals;
}
