// ionic4-hidenav/esm5/lib/hidenav-tabscontent.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive, Host, Self, Optional, ElementRef, ContentChild, forwardRef } from '@angular/core';
import { IonContent } from '@ionic/angular';
import { HidenavService } from './hidenav-service.service';
import { SuperTabs } from '@ionic-super-tabs/angular';
import $ from 'jquery';
var HidenavTabscontentDirective = /** @class */ (function () {
    function HidenavTabscontentDirective(el, contentElem, globals) {
        this.el = el;
        this.contentElem = contentElem;
        this.globals = globals;
    }
    /**
     * @return {?}
     */
    HidenavTabscontentDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.name = this.globals.requestName();
        this.contentElem.nativeElement.setAttribute('hidenav-tabscontent', this.name);
        $('[hidenav-header]', $(this.contentElem.nativeElement).parents().get().find((/**
         * @param {?} itm
         * @return {?}
         */
        function (itm) { return $(itm).find('[hidenav-header]').length; }))).attr('hidenav-header', this.name);
        if (this.name) {
            if (typeof this.globals.data[this.name] == 'undefined' || this.globals.data[this.name] == null)
                this.globals.data[this.name] = [];
            this.globals.data[this.name].tabscontent = this.el;
            this.globals.data[this.name].tabscontentElem = this.contentElem;
            this.globals.initiate(this.name);
            if (this.supertabs) {
                this.supertabs.tabChange.subscribe((/**
                 * @param {?} e
                 * @return {?}
                 */
                function (e) {
                    var e_1, _a;
                    if (e.detail.changed == true) {
                        /** @type {?} */
                        var i = e.detail.index;
                        /** @type {?} */
                        var tabs = _this.contentElem.nativeElement.querySelectorAll('super-tab');
                        /** @type {?} */
                        var results = [];
                        try {
                            for (var tabs_1 = tslib_1.__values(tabs), tabs_1_1 = tabs_1.next(); !tabs_1_1.done; tabs_1_1 = tabs_1.next()) {
                                var tab = tabs_1_1.value;
                                /** @type {?} */
                                var cont = tab.querySelector('ion-content');
                                if (cont.attributes['hidenav-content'])
                                    results.push(cont.attributes['hidenav-content'].nodeValue);
                                else
                                    results.push(null);
                            }
                        }
                        catch (e_1_1) { e_1 = { error: e_1_1 }; }
                        finally {
                            try {
                                if (tabs_1_1 && !tabs_1_1.done && (_a = tabs_1.return)) _a.call(tabs_1);
                            }
                            finally { if (e_1) throw e_1.error; }
                        }
                        if (results[i] != null) {
                            _this.globals.resetTabs(_this.name, results[i]);
                        }
                    }
                }));
            }
        }
    };
    /**
     * @return {?}
     */
    HidenavTabscontentDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.name) {
            delete this.globals.data[this.name].tabscontent;
        }
    };
    HidenavTabscontentDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[hidenav-tabscontent]'
                },] }
    ];
    HidenavTabscontentDirective.ctorParameters = function () { return [
        { type: IonContent, decorators: [{ type: Host }, { type: Self }, { type: Optional }] },
        { type: ElementRef },
        { type: HidenavService }
    ]; };
    HidenavTabscontentDirective.propDecorators = {
        supertabs: [{ type: ContentChild, args: [forwardRef((/**
                     * @return {?}
                     */
                    function () { return SuperTabs; })),] }]
    };
    return HidenavTabscontentDirective;
}());
export { HidenavTabscontentDirective };
if (false) {
    /** @type {?} */
    HidenavTabscontentDirective.prototype.supertabs;
    /** @type {?} */
    HidenavTabscontentDirective.prototype.name;
    /** @type {?} */
    HidenavTabscontentDirective.prototype.el;
    /** @type {?} */
    HidenavTabscontentDirective.prototype.contentElem;
    /**
     * @type {?}
     * @private
     */
    HidenavTabscontentDirective.prototype.globals;
}
