// ionic4-hidenav/esm5/lib/hidenav-sh-tabscontent.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive, Host, Self, Optional, ElementRef, ContentChild } from '@angular/core';
import { IonContent } from '@ionic/angular';
import { SuperTabs } from '@ionic-super-tabs/angular';
import { HidenavShService } from './hidenav-sh-service.service';
import $ from 'jquery';
var HidenavShTabscontentDirective = /** @class */ (function () {
    function HidenavShTabscontentDirective(contentElem, el, globals) {
        this.contentElem = contentElem;
        this.el = el;
        this.globals = globals;
    }
    /**
     * @return {?}
     */
    HidenavShTabscontentDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.name = this.globals.requestName();
        this.contentElem.nativeElement.setAttribute('hidenav-sh-tabscontent', this.name);
        $('hidenav-stretchheader', $(this.contentElem.nativeElement).parents().get().find((/**
         * @param {?} itm
         * @return {?}
         */
        function (itm) { return $(itm).find('[hidenav-stretchheader]').length; }))).attr('hidenav-sh-header', this.name);
        if (this.name) {
            if (typeof this.globals.data[this.name] == 'undefined' || this.globals.data[this.name] == null)
                this.globals.data[this.name] = [];
            this.globals.data[this.name].tabscontent = this.el;
            this.globals.data[this.name].tabscontentElem = this.contentElem;
            this.globals.data[this.name].supertabs = this.supertabs;
            this.globals.initiate(this.name);
            this.supertabs.tabChange.subscribe((/**
             * @param {?} e
             * @return {?}
             */
            function (e) {
                var e_1, _a;
                /** @type {?} */
                var i = e.detail.index;
                /** @type {?} */
                var tabs = _this.contentElem.nativeElement.querySelectorAll('super-tab');
                /** @type {?} */
                var results = [];
                try {
                    for (var tabs_1 = tslib_1.__values(tabs), tabs_1_1 = tabs_1.next(); !tabs_1_1.done; tabs_1_1 = tabs_1.next()) {
                        var tab = tabs_1_1.value;
                        /** @type {?} */
                        var cont = tab.querySelector('ion-content');
                        if (cont != null && typeof cont.attributes['hidenav-sh-content'] != 'undefined' && cont.attributes['hidenav-sh-content'].nodeValue != '')
                            results.push(cont.attributes['hidenav-sh-content'].nodeValue);
                        else
                            results.push(null);
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (tabs_1_1 && !tabs_1_1.done && (_a = tabs_1.return)) _a.call(tabs_1);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
                if (results[i] != null) {
                    _this.globals.resetContent(results[i]);
                }
            }));
        }
    };
    /**
     * @return {?}
     */
    HidenavShTabscontentDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        delete this.globals.data[this.name].tabscontent;
    };
    HidenavShTabscontentDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[hidenav-sh-tabscontent]'
                },] }
    ];
    HidenavShTabscontentDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: IonContent, decorators: [{ type: Host }, { type: Self }, { type: Optional }] },
        { type: HidenavShService }
    ]; };
    HidenavShTabscontentDirective.propDecorators = {
        supertabs: [{ type: ContentChild, args: [SuperTabs,] }]
    };
    return HidenavShTabscontentDirective;
}());
export { HidenavShTabscontentDirective };
if (false) {
    /** @type {?} */
    HidenavShTabscontentDirective.prototype.name;
    /** @type {?} */
    HidenavShTabscontentDirective.prototype.supertabs;
    /** @type {?} */
    HidenavShTabscontentDirective.prototype.contentElem;
    /** @type {?} */
    HidenavShTabscontentDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    HidenavShTabscontentDirective.prototype.globals;
}
