// @ionic-super-tabs/angular/esm5/app-init.js
import * as tslib_1 from "tslib";
import { applyPolyfills, defineCustomElements } from '@ionic-super-tabs/core/loader';
var didInitialize = false;
export function appInit(doc, zone) {
    return function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var win, aelFn;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        win = doc.defaultView;
                        if (!win || didInitialize) {
                            return [2 /*return*/];
                        }
                        didInitialize = true;
                        aelFn = '__zone_symbol__addEventListener' in doc.body
                            ? '__zone_symbol__addEventListener'
                            : 'addEventListener';
                        return [4 /*yield*/, applyPolyfills()];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, defineCustomElements(win, {
                                syncQueue: true,
                                raf: raf,
                                jmp: function (h) { return zone.runOutsideAngular(h); },
                                ael: function (elm, eventName, cb, opts) {
                                    elm[aelFn](eventName, cb, opts);
                                },
                                rel: function (elm, eventName, cb, opts) {
                                    elm.removeEventListener(eventName, cb, opts);
                                },
                            })];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
}
;
export var raf = function (h) {
    if (typeof __zone_symbol__requestAnimationFrame === 'function') {
        return __zone_symbol__requestAnimationFrame(h);
    }
    if (typeof requestAnimationFrame === 'function') {
        return requestAnimationFrame(h);
    }
    return setTimeout(h);
};
