// @ionic-super-tabs/angular/esm5/super-tabs.module.js
import * as tslib_1 from "tslib";
import { CommonModule, DOCUMENT } from '@angular/common';
import { APP_INITIALIZER, NgModule, NgZone } from '@angular/core';
import { appInit } from './app-init';
import { SuperTab, SuperTabButton, SuperTabs, SuperTabsContainer, SuperTabsToolbar } from './directives/proxies';
export var DECLARATIONS = [
    SuperTab,
    SuperTabButton,
    SuperTabs,
    SuperTabsContainer,
    SuperTabsToolbar,
];
var SuperTabsModule = /** @class */ (function () {
    function SuperTabsModule() {
    }
    SuperTabsModule_1 = SuperTabsModule;
    SuperTabsModule.forRoot = function () {
        return {
            ngModule: SuperTabsModule_1,
            providers: [
                {
                    provide: APP_INITIALIZER,
                    useFactory: appInit,
                    multi: true,
                    deps: [DOCUMENT, NgZone],
                },
            ],
        };
    };
    var SuperTabsModule_1;
    SuperTabsModule = SuperTabsModule_1 = tslib_1.__decorate([
        NgModule({
            declarations: DECLARATIONS,
            exports: DECLARATIONS,
            imports: [CommonModule],
        })
    ], SuperTabsModule);
    return SuperTabsModule;
}());
export { SuperTabsModule };
