// @ionic-super-tabs/core/dist/collection/super-tab/super-tab.component.js
import { h } from "@stencil/core";
export class SuperTabComponent {
    componentDidLoad() {
        this.checkIonContent();
    }
    componentDidUpdate() {
        // check for ion-content after update, in case it was dynamically loaded
        this.checkIonContent();
    }
    /**
     * Check if we have an ion-content as a child and update the `noScroll` property
     * if it's not set yet.
     */
    checkIonContent() {
        if (typeof this.noScroll !== 'boolean') {
            const ionContentEl = this.el.querySelector('ion-content');
            if (ionContentEl && ionContentEl.parentElement === this.el) {
                this.noScroll = true;
            }
        }
    }
    /**
     * Returns the root scrollable element
     */
    async getRootScrollableEl() {
        if (!this.noScroll && this.el.scrollHeight > this.el.clientHeight) {
            return this.el;
        }
        const ionContent = this.el.querySelector('ion-content');
        if (ionContent) {
            return ionContent.getScrollElement();
        }
        if (this.noScroll) {
            return null;
        }
        return this.el;
    }
    render() {
        return h("slot", null);
    }
    static get is() { return "super-tab"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() { return {
        "$": ["super-tab.component.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["super-tab.component.css"]
    }; }
    static get properties() { return {
        "noScroll": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": true,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Set this to true to prevent vertical scrolling of this tab. Defaults to `false`.\n\nThis property will automatically be set to true if there is\na direct child element of `ion-content`. To override this\nbehaviour make sure to explicitly set this property to `false`."
            },
            "attribute": "no-scroll",
            "reflect": true
        }
    }; }
    static get methods() { return {
        "getRootScrollableEl": {
            "complexType": {
                "signature": "() => Promise<HTMLElement | null>",
                "parameters": [],
                "references": {
                    "Promise": {
                        "location": "global"
                    },
                    "HTMLElement": {
                        "location": "global"
                    }
                },
                "return": "Promise<HTMLElement | null>"
            },
            "docs": {
                "text": "Returns the root scrollable element",
                "tags": []
            }
        }
    }; }
    static get elementRef() { return "el"; }
}
