// @ionic-super-tabs/angular/esm2015/super-tabs.module.js
import * as tslib_1 from "tslib";
var SuperTabsModule_1;
import { CommonModule, DOCUMENT } from '@angular/common';
import { APP_INITIALIZER, NgModule, NgZone } from '@angular/core';
import { appInit } from './app-init';
import { SuperTab, SuperTabButton, SuperTabs, SuperTabsContainer, SuperTabsToolbar } from './directives/proxies';
export const DECLARATIONS = [
    SuperTab,
    SuperTabButton,
    SuperTabs,
    SuperTabsContainer,
    SuperTabsToolbar,
];
let SuperTabsModule = SuperTabsModule_1 = class SuperTabsModule {
    static forRoot() {
        return {
            ngModule: SuperTabsModule_1,
            providers: [
                {
                    provide: APP_INITIALIZER,
                    useFactory: appInit,
                    multi: true,
                    deps: [DOCUMENT, NgZone],
                },
            ],
        };
    }
};
SuperTabsModule = SuperTabsModule_1 = tslib_1.__decorate([
    NgModule({
        declarations: DECLARATIONS,
        exports: DECLARATIONS,
        imports: [CommonModule],
    })
], SuperTabsModule);
export { SuperTabsModule };
