// ionic4-hidenav/esm2015/lib/hidenav-sh-tabscontent.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Host, Self, Optional, ElementRef, ContentChild } from '@angular/core';
import { IonContent } from '@ionic/angular';
import { SuperTabs } from '@ionic-super-tabs/angular';
import { HidenavShService } from './hidenav-sh-service.service';
import $ from 'jquery';
export class HidenavShTabscontentDirective {
    /**
     * @param {?} contentElem
     * @param {?} el
     * @param {?} globals
     */
    constructor(contentElem, el, globals) {
        this.contentElem = contentElem;
        this.el = el;
        this.globals = globals;
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.name = this.globals.requestName();
        this.contentElem.nativeElement.setAttribute('hidenav-sh-tabscontent', this.name);
        $('hidenav-stretchheader', $(this.contentElem.nativeElement).parents().get().find((/**
         * @param {?} itm
         * @return {?}
         */
        itm => $(itm).find('[hidenav-stretchheader]').length))).attr('hidenav-sh-header', this.name);
        if (this.name) {
            if (typeof this.globals.data[this.name] == 'undefined' || this.globals.data[this.name] == null)
                this.globals.data[this.name] = [];
            this.globals.data[this.name].tabscontent = this.el;
            this.globals.data[this.name].tabscontentElem = this.contentElem;
            this.globals.data[this.name].supertabs = this.supertabs;
            this.globals.initiate(this.name);
            this.supertabs.tabChange.subscribe((/**
             * @param {?} e
             * @return {?}
             */
            e => {
                /** @type {?} */
                let i = e.detail.index;
                /** @type {?} */
                let tabs = this.contentElem.nativeElement.querySelectorAll('super-tab');
                /** @type {?} */
                let results = [];
                for (let tab of tabs) {
                    /** @type {?} */
                    let cont = tab.querySelector('ion-content');
                    if (cont != null && typeof cont.attributes['hidenav-sh-content'] != 'undefined' && cont.attributes['hidenav-sh-content'].nodeValue != '')
                        results.push(cont.attributes['hidenav-sh-content'].nodeValue);
                    else
                        results.push(null);
                }
                if (results[i] != null) {
                    this.globals.resetContent(results[i]);
                }
            }));
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        delete this.globals.data[this.name].tabscontent;
    }
}
HidenavShTabscontentDirective.decorators = [
    { type: Directive, args: [{
                selector: '[hidenav-sh-tabscontent]'
            },] }
];
HidenavShTabscontentDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: IonContent, decorators: [{ type: Host }, { type: Self }, { type: Optional }] },
    { type: HidenavShService }
];
HidenavShTabscontentDirective.propDecorators = {
    supertabs: [{ type: ContentChild, args: [SuperTabs,] }]
};
if (false) {
    /** @type {?} */
    HidenavShTabscontentDirective.prototype.name;
    /** @type {?} */
    HidenavShTabscontentDirective.prototype.supertabs;
    /** @type {?} */
    HidenavShTabscontentDirective.prototype.contentElem;
    /** @type {?} */
    HidenavShTabscontentDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    HidenavShTabscontentDirective.prototype.globals;
}
