// ionic4-hidenav/esm2015/lib/hidenav-tabscontent.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Host, Self, Optional, ElementRef, ContentChild, forwardRef } from '@angular/core';
import { IonContent } from '@ionic/angular';
import { HidenavService } from './hidenav-service.service';
import { SuperTabs } from '@ionic-super-tabs/angular';
import $ from 'jquery';
export class HidenavTabscontentDirective {
    /**
     * @param {?} el
     * @param {?} contentElem
     * @param {?} globals
     */
    constructor(el, contentElem, globals) {
        this.el = el;
        this.contentElem = contentElem;
        this.globals = globals;
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.name = this.globals.requestName();
        this.contentElem.nativeElement.setAttribute('hidenav-tabscontent', this.name);
        $('[hidenav-header]', $(this.contentElem.nativeElement).parents().get().find((/**
         * @param {?} itm
         * @return {?}
         */
        itm => $(itm).find('[hidenav-header]').length))).attr('hidenav-header', this.name);
        if (this.name) {
            if (typeof this.globals.data[this.name] == 'undefined' || this.globals.data[this.name] == null)
                this.globals.data[this.name] = [];
            this.globals.data[this.name].tabscontent = this.el;
            this.globals.data[this.name].tabscontentElem = this.contentElem;
            this.globals.initiate(this.name);
            if (this.supertabs) {
                this.supertabs.tabChange.subscribe((/**
                 * @param {?} e
                 * @return {?}
                 */
                e => {
                    if (e.detail.changed == true) {
                        /** @type {?} */
                        let i = e.detail.index;
                        /** @type {?} */
                        let tabs = this.contentElem.nativeElement.querySelectorAll('super-tab');
                        /** @type {?} */
                        let results = [];
                        for (let tab of tabs) {
                            /** @type {?} */
                            let cont = tab.querySelector('ion-content');
                            if (cont.attributes['hidenav-content'])
                                results.push(cont.attributes['hidenav-content'].nodeValue);
                            else
                                results.push(null);
                        }
                        if (results[i] != null) {
                            this.globals.resetTabs(this.name, results[i]);
                        }
                    }
                }));
            }
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.name) {
            delete this.globals.data[this.name].tabscontent;
        }
    }
}
HidenavTabscontentDirective.decorators = [
    { type: Directive, args: [{
                selector: '[hidenav-tabscontent]'
            },] }
];
HidenavTabscontentDirective.ctorParameters = () => [
    { type: IonContent, decorators: [{ type: Host }, { type: Self }, { type: Optional }] },
    { type: ElementRef },
    { type: HidenavService }
];
HidenavTabscontentDirective.propDecorators = {
    supertabs: [{ type: ContentChild, args: [forwardRef((/**
                 * @return {?}
                 */
                () => SuperTabs)),] }]
};
if (false) {
    /** @type {?} */
    HidenavTabscontentDirective.prototype.supertabs;
    /** @type {?} */
    HidenavTabscontentDirective.prototype.name;
    /** @type {?} */
    HidenavTabscontentDirective.prototype.el;
    /** @type {?} */
    HidenavTabscontentDirective.prototype.contentElem;
    /**
     * @type {?}
     * @private
     */
    HidenavTabscontentDirective.prototype.globals;
}
