// ionic4-hidenav/esm2015/lib/hidenav-header.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef } from '@angular/core';
import { HidenavService } from './hidenav-service.service';
export class HidenavHeaderDirective {
    /**
     * @param {?} el
     * @param {?} globals
     */
    constructor(el, globals) {
        this.el = el;
        this.globals = globals;
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        /** @type {?} */
        const observer = new MutationObserver((/**
         * @param {?} mutations
         * @return {?}
         */
        (mutations) => {
            mutations.forEach((/**
             * @return {?}
             */
            () => {
                if (this.el.nativeElement.getAttribute('hidenav-header').length > 0) {
                    this.name = this.el.nativeElement.getAttribute('hidenav-header');
                    this.start();
                    observer.disconnect();
                }
            }));
        }));
        observer.observe(this.el.nativeElement, {
            attributes: true,
        });
    }
    /**
     * @return {?}
     */
    start() {
        if (this.name) {
            if (typeof this.globals.data[this.name] == 'undefined' || this.globals.data[this.name] == null)
                this.globals.data[this.name] = [];
            if (this.globals.data[this.name].header != null)
                return false;
            this.globals.data[this.name].header = this.el;
            this.globals.initiate(this.name);
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.name) {
            delete this.globals.data[this.name].header;
        }
    }
}
HidenavHeaderDirective.decorators = [
    { type: Directive, args: [{
                selector: '[hidenav-header]'
            },] }
];
HidenavHeaderDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: HidenavService }
];
if (false) {
    /** @type {?} */
    HidenavHeaderDirective.prototype.name;
    /** @type {?} */
    HidenavHeaderDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    HidenavHeaderDirective.prototype.globals;
}
