// ionic4-hidenav/esm2015/lib/hidenav-sh-content.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, Host, Self, Optional, ElementRef } from '@angular/core';
import { IonContent } from '@ionic/angular';
import { HidenavShService } from './hidenav-sh-service.service';
import $ from 'jquery';
export class HidenavShContentDirective {
    /**
     * @param {?} el
     * @param {?} contentElem
     * @param {?} globals
     */
    constructor(el, contentElem, globals) {
        this.el = el;
        this.contentElem = contentElem;
        this.globals = globals;
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        if (!this.contentElem.nativeElement.hasAttribute('hidenav-tabspage')) {
            this.name = this.globals.requestName();
            $(this.contentElem.nativeElement).attr('hidenav-sh-content', this.name);
            $('hidenav-stretchheader', $(this.contentElem.nativeElement).parents().get().find((/**
             * @param {?} itm
             * @return {?}
             */
            itm => $(itm).find('[hidenav-header]').length))).attr('hidenav-sh-header', this.name);
            this.start();
        }
        else {
            /** @type {?} */
            let counter = 0;
            /** @type {?} */
            let int = setInterval((/**
             * @return {?}
             */
            () => {
                /** @type {?} */
                let x = $(this.contentElem.nativeElement).closest('[hidenav-sh-tabscontent]').attr('hidenav-sh-tabscontent');
                counter++;
                if (x && x.length > 0) {
                    this.parent = $(this.contentElem.nativeElement).closest('[hidenav-sh-tabscontent]').attr('hidenav-sh-tabscontent');
                    this.name = this.globals.requestTabName(this.parent);
                    $(this.contentElem.nativeElement).attr('hidenav-sh-content', this.name);
                    $(this.contentElem.nativeElement).attr('hidenav-tabspage', this.parent);
                    this.start();
                    clearInterval(int);
                }
                else if (counter > 50) {
                    clearInterval(int);
                }
            }), 50);
        }
    }
    /**
     * @return {?}
     */
    start() {
        if (this.name) {
            if (typeof this.globals.data[this.name] == 'undefined' || this.globals.data[this.name] == null)
                this.globals.data[this.name] = [];
            this.globals.data[this.name].content = this.el;
            this.globals.data[this.name].contentEl = this.contentElem;
            if (this.parent)
                this.globals.data[this.name].parent = this.parent;
            this.globals.initiate(this.name);
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.name) {
            delete this.globals.data[this.name].content;
        }
    }
}
HidenavShContentDirective.decorators = [
    { type: Directive, args: [{
                selector: '[hidenav-sh-content]'
            },] }
];
HidenavShContentDirective.ctorParameters = () => [
    { type: IonContent, decorators: [{ type: Host }, { type: Self }, { type: Optional }] },
    { type: ElementRef },
    { type: HidenavShService }
];
HidenavShContentDirective.propDecorators = {
    hntb: [{ type: Input, args: ['hidenav-tabspage',] }]
};
if (false) {
    /** @type {?} */
    HidenavShContentDirective.prototype.parent;
    /** @type {?} */
    HidenavShContentDirective.prototype.name;
    /** @type {?} */
    HidenavShContentDirective.prototype.hntb;
    /** @type {?} */
    HidenavShContentDirective.prototype.el;
    /** @type {?} */
    HidenavShContentDirective.prototype.contentElem;
    /**
     * @type {?}
     * @private
     */
    HidenavShContentDirective.prototype.globals;
}
